### ISSUE TITLE
Discussion: websearch_to_tsquery

### ISSUE THREAD

**Issue by data-envoy** [dekimir/postgres-searchbox#7]
I've notice that there is a subtle difference between `websearch_to_tsquery` [docs](https://www.postgresql.org/docs/current/textsearch-controls.html) & how most search engines work.

Say the database contians 2 entries:

```js
const data = [
  {
    "id": 1,
    "name": "Awesome Concrete Towels",
    "description": "The Football Is Good For Training And Recreational Purposes"
  },
  {
    "id": 2,
    "name": "Great Concrete Towels"
    "description": "The Football Is Good For Training And Recreational Purposes",
  }
];
```

A search for `awesome football` will be parsed `websearch_to_tsquery(awesome football)` to 'awesome' & 'football'. 

It only returns entry 1, where I think it should return both.

I tried to make a search return both entries by doing `awesome OR football` but that didn't work. It excludes entry 1 because it matches both terms, so it fails the OR condition.

---

Having the desired result being both entries, I think we need to do some additonal work.

**Comment by dekimir** [comment]
> It only returns entry 1, where I think it should return both.

That would depend on what we call "the match mode".  See [this link](https://docs.oracle.com/cd/F25151_01/mdex.11-3-2/EndecaMDEXDev/html/cbdv_searchmodes_list.xmlconcept_01FC5849F8D44A3CB4DC1ADD32F0BEFD.html) for explanation.  I believe Algolia's default behaviour is to [require all terms to match](https://www.algolia.com/doc/guides/managing-results/relevance-overview/in-depth/ranking-criteria/#words-if-applicable), which is exactly what Postgres is doing here.

> I tried to make a search return both entries by doing `awesome OR football` but that didn't work. It excludes entry 1 because it matches both terms, so it fails the OR condition.

Perhaps you should try with `|` instead of `OR`?  When I try that, I get this:
```sql
dejan> select 'football is good' @@ to_tsquery('awesome');
+----------+
| ?column? |
|----------|
| False    |
+----------+
SELECT 1
Time: 0.003s
dejan> select 'football is good' @@ to_tsquery('awesome | football');
+----------+
| ?column? |
|----------|
| True     |
+----------+
SELECT 1
Time: 0.003s
```

**Comment by data-envoy** [comment]
Oh yes, you are right about Algolias eault being `AND`. I would really love to look into ranking by optionalWords like they do.

I tried your suggestion using `|`, it works for `to_tsquery` but not for `websearch_to_tsquery` ( I can't seem to find an equivalent )

**Comment by dekimir** [comment]
`OR` works for me in websearch_to_tsquery:
```sql
dejan> select 'football is good' @@ websearch_to_tsquery('awesome or football');
+----------+
| ?column? |
|----------|
| True     |
+----------+
SELECT 1
Time: 0.003s
```

**Comment by dekimir** [comment]
> Oh yes, you are right about Algolias eault being `AND`. I would really love to look into ranking by optionalWords like they do.

Yes, we'll get to that.  But it'll likely require us implementing our own query parser (similar to https://github.com/dekimir/postgres-searchbox/issues/1#issuecomment-1505831937) or even a Postgres extension.  It's doable, but it'll take some work.
